src/protocol/main.mo

```diff
@@ -1,5 +1,117 @@
-actor Protocol {
-  public func greet(name : Text) : async Text {
-    return "Hello, " # name # "!";
+import Token "../mint_token/token";
+import Cycles "mo:base/ExperimentalCycles";
+import Nat "mo:base/Nat";
+import Option "mo:base/Option";
+import Array "mo:base/Array";
+import Principal "mo:base/Principal";
+import HashMap "mo:base/HashMap";
+import Prim "mo:prim";
+import P "./position";
+import Error "mo:base/Error";
+
+actor Minter {
+  // FOR DIP20 INTERFACE
+  type TxReceipt = {
+      #Ok: Nat;
+      #Err: {
+          #InsufficientAllowance;
+          #InsufficientBalance;
+          #ErrorOperationStyle;
+          #Unauthorized;
+          #LedgerTrap;
+          #ErrorTo;
+          #Other: Text;
+          #BlockUsed;
+          #AmountTooSmall;
+      };
   };
+  type DIP20 = actor {
+     transferFrom(from: Principal, to: Principal, value: Nat): async TxReceipt;
+     transfer(to: Principal, value: Nat): async TxReceipt;
+};
+  
+  var collateralActor: ?DIP20 = null;
+  var usbActor: ?Token.UsbToken = null;
+  var lastPositionId : Nat = 0;
+  var collateralPrice = 0;
+  var minRisk = 15;
+  let positionMap = HashMap.HashMap<Nat, P.Position>(
+    0,
+    Nat.equal,
+    func(x)   { Prim.natToNat32 x }
+  );
+  let accountPositions = HashMap.HashMap<Principal, [Nat]>(
+    0,
+    Principal.equal,
+    Principal.hash
+  );
+
+  public func init(collateralActorText: Text): async () {
+    // Do not allow to call init more then once
+    assert Option.isNull(usbActor);
+    Cycles.add(1_000_000_000_000);
+
+    usbActor := ?(await Token.UsbToken("", "Decentralized USD", "USB", 18, 0, 100));
+    collateralActor := ?(actor(collateralActorText));
+  };
+
+  public query func getTokenPrincipal(): async Principal {
+    switch (usbActor) {
+      case null {
+        throw Error.reject("Contract is not initialized.");
+      };
+      case (?usbActor) {
+        return Principal.fromActor(usbActor);
+      };
+    };
+  };
+
+  public func setCollateralPrice(newPrice: Nat): async () {
```

Comment:
assert oracle?

Reply:
No assert right now
Because, we don't have oracle keys, and it is not so important for mvp